bazel-build.el

```diff
@@ -26,24 +26,62 @@
 
 (defun bazel-build (target)
   "Build a Bazel TARGET."
-  (interactive "sbazel build ")
+  (interactive (list (bazel-build--read-target "bazel build ")))
   (bazel-build--run-bazel-command "build" target))
 
 (defun bazel-run (target)
   "Build and run a Bazel TARGET."
-  (interactive "sbazel run ")
+  (interactive (list (bazel-build--read-target "bazel run ")))
   (bazel-build--run-bazel-command "run" target))
 
 (defun bazel-test (target)
   "Build and run a Bazel test TARGET."
-  (interactive "sbazel test ")
+  (interactive (list (bazel-build--read-target "bazel test ")))
   (bazel-build--run-bazel-command "test" target))
 
 (defun bazel-build--run-bazel-command (command target)
   "Run Bazel tool with given COMMAND, e.g. build or run, on the given TARGET."
   (compile
    (mapconcat #'shell-quote-argument (list "bazel" command target) " ")))
 
+(defun bazel-build--read-target (prompt)
+  "Read a Bazel build target from the minibuffer.  PROMPT is a read-only prompt."
+  (let* ((file-name (buffer-file-name))
+         (workspace-root
+          (or (bazel-build--find-workspace-root file-name)
+              (user-error "Not in a Bazel workspace. No WORKSPACE file found.")))
+         (package-name
+          (or (bazel-build--extract-package-name file-name workspace-root)
+              (user-error "Not in a Bazel package. No BUILD file found.")))
+         (initial-input (concat "//" package-name)))
+    (read-string prompt initial-input)))
+
+(defun bazel-build--find-workspace-root (file-name)
+  "Find the root of the Bazel workspace containing FILE-NAME.
+If FILE-NAME is not in a Bazel workspace, return nil."
+  (let ((workspace-root (locate-dominating-file file-name "WORKSPACE")))
+    (when workspace-root workspace-root)))
+
+(defun bazel-build--extract-package-name (file-name workspace-root)
+  "Return the nearest Bazel package for FILE-NAME under WORKSPACE-ROOT.
+If FILE-NAME is not in a Bazel package, return nil."
+  (let* ((build-file-directory
+          (cl-some (lambda (build-name)
+                     (locate-dominating-file file-name build-name))
+                   '("BUILD.bazel" "BUILD")))
+         (package-name
+          (when build-file-directory
+            (cond ((string= workspace-root build-file-directory) "")
+                  ((string-prefix-p workspace-root build-file-directory)
+                   (file-relative-name build-file-directory workspace-root))
+                  (t nil)))))
```

Comment:
Cases are:
  a) BUILD file in the same directory as the WORKSPACE file. This was formerly returning the relative path of "." causing the suggested input to be "//.", but this special case turns that to "//"
  b) BUILD file is in a subdirectory of the workspace root.
  c) BUILD file is found in some parent directory of the WORKSPACE file. This is a nonsense situation that wasn't handled properly before either but now is.

By returning a relative directory from bazel-build--find-workspace-root, we avoid the need to call expand-file-name on build-file-directory, which in turn avoids the need to handle the nil case before doing so.

Reply:
I think this shouldn't use `string-prefix-p`, as plain string functions tend to not work well with filenames. (What if there's a `/./` part in the filename?)

Can you use `file-equal-p` and `file-in-directory-p` instead?

Reply:
Done.